File: tribonacci_sequence/tribonacci_sequence.java
import java.util.stream.DoubleStream;
import java.util.Arrays;

public class Xbonacci {
    private DoubleStream trib(double[] signature) {
        return DoubleStream.generate(
            () -> { 
                double res = signature[0] + signature[1] + signature[2]; 
                signature[0] = signature[1];
                signature[1] = signature[2];
                signature[2] = res;
                return res;
            });
    }

    public double[] tribonacci(double[] s, int n) {
        return DoubleStream.concat(Arrays.stream(s), trib((s))).limit(n).toArray();
    }
}

File: tribonacci_sequence/tribonacci_sequence.go
package kata

func Trib(signature [3]float64) chan float64 {
	ch := make(chan float64)
	go func() {
		for true {
			k := signature[0] + signature[1] + signature[2]
			signature[0] = signature[1]
			signature[1] = signature[2]
			signature[2] = k
			ch <- k
		}
	}()
	return ch
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func Tribonacci(signature [3]float64, n int) []float64 {
	res := signature[:min(3, n)]
	trib_gen := Trib(signature)
	for i := 3; i < n; i++ {
		res = append(res, <-trib_gen)
	}
	return res
}
